feat: omada-controller helm chart

gca3020 opened the pull request:
# Description

After the brief discussion in #653, this PR adds a helm chart to the repository. This helm chart is based on my own personal usage, but I have attempted to update it to support additional functionality, such as injecting certificates directly into the container using `extraVolumes` as well as supporting standard K8s best practices like allowing additional labels and annotations.

Out of the box, it is intended to be used with dynamic volume provisioning, as well as a LoadBalancer and Ingress. It was tested with MetalLB and Nginx ingress, and works with both, but additional ingress and LoadBalancer providers should be supported as well.

By default, it uses the (relatively new) rootless functionality of the underlying image. However, it also supports disabling this and running as root by updating the `config.rootless` setting in the Values file. Older Kubernetes versions (e.g. <1.19) do not support things like the `seccompProfile` configuration, so this was added to make sure that this chart is compatible with older versions of K8s as well as newer.

mbentley commented on helm/omada-controller/templates/statefulset.yaml:
```diff
@@ -0,0 +1,214 @@
+  labels:
+    {{- include "omada-controller.labels" . | nindent 4 }}
+spec:
+  replicas: 1
```
Just a nit but if we are defining this in `values.yaml`, maybe we should use it here otherwise we should remove the value if we are not going to reference it or just maybe have it as a comment in the values.yaml stating that it's hard coded for a good reason.

gca3020 replied:
Good catch. As I mentioned in the comment in the values, there's not a lot of reason to run multiple replicas, when it's generally easier to install it as two different releases or in two different namespaces, but I can understand that people might want to, so I left the option in place from the default helm template.

mbentley reviewed (dismissed):
I added a few comments - probably a few things that might need tweaking although I am no Helm expert so happy for any push back if any of my comments do not make sense 😄 

I can add a README as I will probably add some automation to publish Helm charts to the Docker Hub repo once I have that set up.

gca3020 commented on helm/omada-controller/templates/statefulset.yaml:
```diff
@@ -0,0 +1,214 @@
+      spec:
+        accessModes: 
+          {{- toYaml .Values.persistence.data.accessModes | nindent 10 }}
+        storageClassName: {{ .Values.persistence.data.storageClass }}
```
Again, good catch here. I only have one StorageClass in my test cluster, and it's the default, so I didn't actually test overriding it with a value, because the empty string works fine for the default SC.

gca3020 commented:
Thanks for the feedback, I'm addressing all the comments and will push an update shortly. I also found another issue where the Service is invalid when deploying from a fresh state because the portal and management http ports are the same. It worked during an upgrade so I missed it. 

As for the name, that's fine. I wasn't sure if you were planning on pushing to an OCI container registry, or using the github chart releaser action and github pages, which is what I've done in the past. Would you like me to rename the Chart.yaml and folder it's in, or would you like me to also update all the _helpers.yaml templates (not strictly necessary). 

It's also worth noting that once installed, if the chart name is `omada-controller-helm` and you install it with a release name of `omada-controller` you'll end up with the objects named `omada-controller-omada-controller-helm` which gets a bit verbose. 

I am not super familiar with charts in OCI registries, but I thought you could have subfolders, so maybe you could have `mbentley/charts/omada-controller` or `mbentley/helm/omada-controller` instead?

I found this issue describing exactly what you're talking about, and recommends adding a `charts/` subdirectory. I'm not sure if Dockerhub allows this or not. https://github.com/helm/helm-www/issues/1717

gca3020 commented on helm/omada-controller/Chart.yaml:
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
+name: omada-controller
```
Just wanted to post here and ensure you saw my comment @mbentley. No rush on this, just checking in to see what your thoughts were about naming it.

gca3020 commented:
Ok @mbentley, I believe I have a good workaround here, that **mostly** adheres to what I've found for helm best practices. 

Commit [d3532a0](https://github.com/mbentley/docker-omada-controller/pull/668/commits/d3532a067fe3cbb8276cb58f3d3784342f9271a9) has the core of it:
- Rename the chart to `omada-controller-helm`
- Trim the "-helm" suffix in the helpers, *except* where it explicitly refers to the chart by name (e.g. `helm.sh/chart: omada-controller-helm-1.0.0`, to avoid confusion.
